src/Components/Endpoints/src/DependencyInjection/RazorComponentsServiceCollectionExtensions.cs

```diff
@@ -54,6 +54,7 @@ public static IRazorComponentsBuilder AddRazorComponents(this IServiceCollection
         services.TryAddSingleton<WebAssemblyComponentSerializer>();
         services.TryAddScoped<EndpointHtmlRenderer>();
         services.TryAddScoped<IComponentPrerenderer>(services => services.GetRequiredService<EndpointHtmlRenderer>());
+        services.TryAddScoped<ISerializationModeHandler>(services => services.GetRequiredService<EndpointHtmlRenderer>());
```

Comment:
This is where the issue is. There is a hidden circular dependency here.

Renderer -> PersistentComponentStateManager -> Renderer(ISerializationModeHandler).

The thing that initiates the persistence process (the renderer) calls the `ComponentPersistenceManager` and has access to the renderer, which in turn can use that to determine the correct render mode https://github.com/dotnet/aspnetcore/blob/8eaf4b51f73ae2b0ed079e4f8253afb53e96b703/src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs#L50